source/extensions/http/cache/file_system_http_cache/cache_file_fixed_block.h

```diff
@@ -0,0 +1,74 @@
+#pragma once
+
+#include <sys/types.h>
+
+#include <cstddef>
+#include <cstdint>
+
+#include "absl/strings/string_view.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace HttpFilters {
+namespace Cache {
+namespace FileSystemHttpCache {
+
+// This represents a minimal header block on the cache entry file; it uses a struct that
+// is required to be packed, and explicit network byte order, to ensure compatibility.
+class CacheFileFixedBlock {
+public:
+  CacheFileFixedBlock();
+  void populateFromStringView(absl::string_view s);
+  static size_t size() { return sizeof(contents_); }
+  // fileId is a fixed value used to identify that this is a cache file.
+  uint32_t fileId() const { return getUint32(contents_.file_id_); }
+  // cacheVersionId is a value that should be consistent between versions of the file
+  // cache implementation. Changing version in code will invalidate all cache
+  // entries where the version ID does not match.
+  uint32_t cacheVersionId() const { return getUint32(contents_.cache_version_id_); }
+  // Size of the serialized proto message capturing headers and metadata.
+  size_t headerSize() const { return getUint32(contents_.header_size_); }
+  // Size of the serialized proto message capturing trailers.
+  size_t trailerSize() const { return getUint32(contents_.trailer_size_); }
+  // Size of the http body of the cache entry.
+  size_t bodySize() const { return getUint64(contents_.body_size_); }
+  void setFileId(uint32_t id) { setUint32(contents_.file_id_, id); }
+  void setCacheVersionId(uint32_t id) { setUint32(contents_.cache_version_id_, id); }
+  void setHeadersSize(size_t sz) { setUint32(contents_.header_size_, sz); }
+  void setTrailersSize(size_t sz) { setUint32(contents_.trailer_size_, sz); }
+  void setBodySize(size_t sz) { setUint64(contents_.body_size_, sz); }
+  static size_t offsetToHeaders() { return size(); }
+  size_t offsetToBody() const { return offsetToHeaders() + headerSize(); }
+  size_t offsetToTrailers() const { return offsetToBody() + bodySize(); }
+  absl::string_view stringView() const { return {contents_.as_str_, size()}; }
```

Comment:
This access to contents_.as_str_ is undefined behavior because as_str_ is not the active member of the union. If you ditch the union and have contents_ directly hold the uints, this can be replaced with `return { reinterpret_cast<char*>(&contents_), size()}; }` (one of the rare uses of reinterpret_cast that doesn't lead to UB).